### PR TITLE
SOLR-17156: introducing RandomNoReverseMergePolicyFactory

### PR DESCRIPTION
**Pull request by mkhludnev**
healing BJQParserTest

https://issues.apache.org/jira/browse/SOLR-17156

### CODE REVIEW

**Comment by mkhludnev** [comment]
It needs to be spread across other block join tests as well.

**Review by janhoy — APPROVED** [review]
Cool!

**Comment by mkhludnev** [comment]
The scope has blown up. Are there other tests rely on block alignment?

File: solr/test-framework/src/java/org/apache/solr/util/RandomNoReverseMergePolicyFactory.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * limitations under the License.
+ */
+package org.apache.solr.util;
+
+import org.apache.lucene.index.MergePolicy;
+import org.apache.solr.index.MergePolicyFactory;
+import org.apache.solr.index.MergePolicyFactoryArgs;
+
+/** A {@link MergePolicyFactory} for {@link RandomMergePolicy}. */
+public final class RandomNoReverseMergePolicyFactory extends MergePolicyFactory {
```
Comment (dsmiley) [inline solr/test-framework/src/java/org/apache/solr/util/RandomNoReverseMergePolicyFactory.java:24]: I don't understand how this prevents "ReverseMergePolicy"

Comment (mkhludnev) [reply solr/test-framework/src/java/org/apache/solr/util/RandomNoReverseMergePolicyFactory.java:24]: Here we go 
https://github.com/apache/solr/pull/2289/files/d24bfff1fe1aaa11a666c8f7188895ce23eed128#diff-681b5f32ae2e1f988d11d5112305b5c71e08b3c3abee5879db27a9b91d63b398R46

Comment (mkhludnev) [reply solr/test-framework/src/java/org/apache/solr/util/RandomNoReverseMergePolicyFactory.java:24]: @dsmiley thanks for the guidance. I'll check it out.

**Comment by mkhludnev** [comment]
@dsmiley thanks for the guidance. I'll check it out.

**Comment by mkhludnev** [comment]
> SystemPropertiesRestoreRule instead. See its constructor that takes the key and value. 

I'm ok with this proposal @dsmiley. But I can't find any usage of two args constructor, so far all tests use `systemSetPropertySolrTestsMergePolicyFactory`. As far as I understand `SystemPropertiesRestoreRule` rule already applied and restores props thus, I don't need to bother about restoring in `@AfterClass` - ok. 
Then if I use this rule, I need to copy-paste prop name from `systemSetPropertySolrTestsMergePolicyFactory` that doesn't seem good for me. if this bi-arg constructor works, should I use `RuleChain` or just `@Rule`?     

Note: I changed `SolrTestCase` just to fix its' subclasses which don't inherit `SolrTestCase4J`.

**Review by dsmiley — APPROVED** [review]
Nice

**Comment by mkhludnev** [comment]
I think it's test only change, thus it doesn't deserves CHANGES entry

**Pull request merged**